File: python solutions/binary_tree_paths.py
# Definition for a binary tree node.
from typing import List, Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def binaryTreePaths(self, root: Optional[TreeNode]) -> List[str]:
        res = []

        def helper(node, curr):
            if not node:
                return
            if not node.left and not node.right:
                res.append(curr + [str(node.val)])
            else:
                helper(node.left, curr + [str(node.val)])
                helper(node.right, curr + [str(node.val)])
        
        helper(node=root, curr=[])
        return [*map('->'.join, res)]

File: ruby solutions/binary_tree_paths.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {String[]}
def binary_tree_paths(root)
    @res = []
    helper(node=root, curr=[])
    
    return @res.map{ |a| a.join("->") }
end

def helper(node, curr)
    return if !node
    if !node.left && !node.right
        @res.append(curr + [node.val.to_s])
    else
        helper(node.left, curr + [node.val.to_s])
        helper(node.right, curr + [node.val.to_s])
    end
end